### PR TITLE
Use NavigationAgent3D in 3d/navigation demo

### PR DESCRIPTION
**Pull request by idbrii**
This is a port of #806 to Godot 4.

I put type hints in a separate commit, but I can squash with other code changes if you prefer.

Add NavigationAgent3D to robot and add a script on the robot to use it.

Remove DrawPath and replace with dynamically created Line3D that makes
it easier for users to copy to their projects (just copy the code
instead of copying multiple parts out of a scene).

There are several auto applied changes to the scene that I'm not sure
about. I'm saving in Godot v4.1.1.stable.official [bd6af8e0e] and it
appears that it no longer uses uids?

While I made these changes on Godot 4.1, I did not commit the changes to project.godot. It gets reordered and sets `config/features=PackedStringArray("4.1")`. I'm not sure if demos should be on the latest stable?

I ran into an update_polygons error message, but the only way I could find to fix was to revert the nav mesh cell size (I don't know where to set the nav map cell size or why it doesn't match). I don't see a difference in behaviour.

# Lighting 
I also changed the lighting to actually illuminate the ground:

**Before**
before

**After**
after

### CODE REVIEW

File: 3d/navigation/character.gd [hunk 1]
```diff
@@ -0,0 +1,45 @@
+extends Marker3D
+
+const Line3D = preload("res://line3d.gd")
```
Comment (AThousandShips) [inline 3d/navigation/character.gd:3]:
```suggestion

const Line3D = preload("res://line3d.gd")
```

File: 3d/navigation/character.gd [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		var start_position := global_transform.origin
+		var optimize := true
+		var navigation_map := get_world_3d().get_navigation_map()
+		var path := NavigationServer3D.map_get_path(
+				navigation_map,
+				start_position,
+				target_position,
+				optimize)
+		_nav_path_line.draw_path(path)
+
```
Comment (AThousandShips) [inline 3d/navigation/character.gd:45]:
```suggestion
```

File: 3d/navigation/character.gd [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	# Make the robot look at the direction we're traveling.
+	# Clamp y to 0 so the robot only looks left and right, not up/down.
+	offset.y = 0
+	look_at(global_position + offset, Vector3.UP)
+
+
+func set_target_position(target_position: Vector3):
+	_nav_agent.set_target_position(target_position)
+	# get a full navigation path with the NavigationServer API
```
Comment (AThousandShips) [inline 3d/navigation/character.gd:34]:
```suggestion
	# Get a full navigation path with the NavigationServer API.
```

File: 3d/navigation/navmesh.gd [hunk 4]
```diff
@@ -1,84 +1,31 @@
 extends Node3D
 
-const SPEED := 10.0
+const Character = preload("res://character.gd")
```
Comment (AThousandShips) [inline 3d/navigation/navmesh.gd:3]:
```suggestion

const Character = preload("res://character.gd")
```

File: 3d/navigation/navmesh.gd [hunk 5]
```diff
@@ -1,84 +1,31 @@
...
-		var target_point := NavigationServer3D.map_get_closest_point_to_segment(map, from, to)
-
-		# Set the path between the robot's current location and our target.
-		path = NavigationServer3D.map_get_path(map, robot.position, target_point, true)
-
-		if show_path:
-			draw_path(path)
-
-	elif event is InputEventMouseMotion:
+		# get closest point on navmesh for the current mouse cursor position
```
Comment (AThousandShips) [inline 3d/navigation/navmesh.gd:13]:
```suggestion
		# Get closest point on navmesh for the current mouse cursor position.
```

File: 3d/navigation/navmesh.gd [hunk 6]
```diff
@@ -1,84 +1,31 @@
...
+		_robot.set_target_position(closest_point_on_navmesh)
+
+	if event is InputEventMouseMotion:
 		if event.button_mask & (MOUSE_BUTTON_MASK_MIDDLE + MOUSE_BUTTON_MASK_RIGHT):
-			cam_rotation += event.relative.x * 0.005
-			$CameraBase.set_rotation(Vector3(0, cam_rotation, 0))
-
+			_cam_rotation += event.relative.x * 0.005
+			$CameraBase.set_rotation(Vector3.UP * _cam_rotation)
+			print("Camera Rotation: ", _cam_rotation)
```
Comment (AThousandShips) [inline 3d/navigation/navmesh.gd:31]:
```suggestion
```

File: 3d/navigation/navmesh.gd [hunk 7]
```diff
@@ -1,84 +1,31 @@
...
+		var camera_ray_end := camera_ray_start + _camera.project_ray_normal(mouse_cursor_position) * camera_ray_length
+
+		var closest_point_on_navmesh := NavigationServer3D.map_get_closest_point_to_segment(
+			get_world_3d().navigation_map,
+			camera_ray_start,
+			camera_ray_end
+			)
+		_robot.set_target_position(closest_point_on_navmesh)
+
+	if event is InputEventMouseMotion:
```
Comment (AThousandShips) [inline 3d/navigation/navmesh.gd:26]:
```suggestion
	elif event is InputEventMouseMotion:
```

**Comment by idbrii** [comment]
Folded those review changes into the corresponding commits.

**Comment by Calinou** [comment]
> While I made these changes on Godot 4.1, I did not commit the changes to project.godot. It gets reordered and sets config/features=PackedStringArray("4.1"). I'm not sure if demos should be on the latest stable?

It's fine to commit that part. We want the `master` branch to run on Godot 4.1.x as well as 4.2 betas.

**Review by Calinou — COMMENTED** [review]
Tested locally on Godot 4.1.2, it works as expected.

**Review by Calinou — APPROVED** [review]
Thanks!

**Pull request merged**